pkg/agent/informant.go

```diff
@@ -199,16 +199,26 @@ func NewInformantServer(
 
 		// we need to spawn these in separate threads so the caller doesn't block while holding
 		// runner.lock
-		runner.spawnBackgroundWorker(context.TODO(), shutdownName, func(c context.Context) {
-			if err := httpServer.Shutdown(c); err != nil {
+		runner.spawnBackgroundWorker(ctx, shutdownName, func(context.Context) {
+			// we want shutdown to (potentially) live longer than the request which
+			// made it, but having a timeout is still good.
+			ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+			defer cancel()
+
+			if err := httpServer.Shutdown(ctx); err != nil {
 				runner.logger.Warningf("Error shutting down InformantServer: %s", err)
 			}
 		})
 		if server.madeContact {
 			// only unregister the server if we could have plausibly contacted the informant
 			unregisterName := fmt.Sprintf("InformantServer unregister (%s)", server.desc.AgentID)
-			runner.spawnBackgroundWorker(context.TODO(), unregisterName, func(c context.Context) {
-				if err := server.unregisterFromInformant(c); err != nil {
+			runner.spawnBackgroundWorker(ctx, unregisterName, func(context.Context) {
+				// we want shutdown to (potentially) live longer than the request which
+				// made it, but having a timeout is still good.
+				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+				defer cancel()
+
+				if err := server.unregisterFromInformant(ctx); err != nil {
```

Comment:
`unregisterFromInformant` already uses a timeout on the request itself; do we need another one?

Reply:
I think the fact that `doInformantRequest` takes a timeout *and* a context is a bit of a weird API, and I kind of planned to pull that apart in a later PR but I don't feel rushed about that.. 

it's plausible that we could just pass the the enclosing context to the unregister call, and not worry about it during shutdown, but.